### DISCUSSION TITLE
macOS: font-thicken behaviour

### DISCUSSION THREAD

**Discussion by vitvly** [ghostty-org/ghostty#3492] [Q&A] [answered]
Hi!

Thanks for great software. I've been eagerly waiting for it :)

I have a question: how can i make font rendering completely identical to Terminal.app? Consider below:
<img width="700" alt="Image" src="https://github.com/user-attachments/assets/c16e49db-aa85-4e62-9ecf-0087fb729f37" />

The image above is Terminal.app, below is Ghostty with `font-thicken` setting. Same font and size (Menlo 12). Ghostty's version is a bit heavier (and horizontally wider).

**Comment by jaredh159** [comment]
possibly related: Kitty supports a numeric value for `font_thicken`, I had mine turned to `0.2`, but it seems to be just a bool in Ghostty and turning it on is _too thick_ for me as well, whereas I feel like without it it's too thin.

love ghostty so far and really appreciate all the work!

**Reply by vitvly** [reply]
Yes, i also recall using Kitty’s font thickening feature, and landing on somewhere between 0.2 and 0.3. However, i was not ever able to figure out the set of options that would produce fonts exactly similar to Terminal.app and TextEdit.

I think it would be great if default font settings in Ghostty were as close to these two apps as possible. However, i understand that font rendering is extremely complicated.

**Comment by sampotts** [comment]
My 2c: rather than having to rely on `font_thicken`, can we have some control over the antialiasing on macOS? I feel like it's too strong at the moment, rendering most fonts too thin. Then you have to resort to `font_thicken` which in turn makes the type not quite as sharp as it should be.

**Comment by stancl** [comment]
_Originally wanted to submit a separate discussion but this is likely related enough_

I'm using a [particular font](https://practicaltypography.com/triplicate.html) and encountering some rendering issues. However, similar behavior happens in some other editors so I'm not sure if this is specifically a bug. It may instead be a feature request.

Essentially, the font renders great in Kitty. See [light mode](https://share.cleanshot.com/9xgSWsyP) and [dark mode](https://share.cleanshot.com/xM2HGbvw).

In Ghostty it renders like this:
- [Light mode, no font-thicken](https://share.cleanshot.com/MZqqM99S) looks alright-ish
- [Light mode, font-thicken](https://share.cleanshot.com/dkzRLHXy) looks very bad
- [Dark mode, no font-thicken](https://share.cleanshot.com/BcNWJMNg) looks bad
- [Dark mode, font-thicken](https://share.cleanshot.com/RX90L4Ms) looks best

It's as if `font-thicken` makes dark mode look almost exactly like Kitty, but makes light mode look worse. I see the other discussion I linked mentioned adding a numeric value for the `font-thicken` setting — could that be it, or is there more going on in Kitty, like additional font rendering logic that tries to adjust text to its background perhaps? **A quick search found [this](https://github.com/kovidgoyal/kitty/blob/dac6f88453ed97097f543c1db4181e70480eb6a5/docs/changelog.rst#0280-2023-04-15)** (https://github.com/kovidgoyal/kitty/pull/5969).

For reference, the font also looks kinda ugly in Zed ([dark mode](https://share.cleanshot.com/9XpCg7Lb), [light mode](https://share.cleanshot.com/7W8SYyFn)) but a little better I'd say. So with my limited (essentially nonexistent) knowledge of font rendering, I'm inclined to think Kitty puts in extra work into rendering fonts nicely, rather than Ghostty doing something wrong?

**Comment by qoobes** [comment]
Fonts are pretty awful for me as well compared to kitty. Using the firacode mono nerd font zoomed out to 10px fontsize

**Comment by vitvly** [comment]
Also: if i recall correctly, in the Changelog podcast about Ghostty (https://changelog.com/podcast/622) it was said that TextEdit is used as a standard to compare against when implementing font rendering on macOS.

I think it's an absolutely correct idea. Given that Ghostty aims at native look & feel, it's important to have a standard macOS font look *by default*, so that the user doesn't have to tweak any knobs, like `font-thicken` etc. Or, at least, have a documented away to achieve that default look.

**Comment by MrConnorKenway** [comment]
With `font-thicken = true`, some PUA symbols becomes distorted compared with kitty:

### kitty:
<img width="100" alt="image" src="https://github.com/user-attachments/assets/2caa0864-e42e-499a-8e29-262a8d871def" />

### ghostty w/ font thicken:
<img width="200" alt="image" src="https://github.com/user-attachments/assets/b0a49a95-7d38-4f17-ac00-85961624be51" />

### ghostty w/o font thicken:
<img width="100" alt="image" src="https://github.com/user-attachments/assets/27140f0b-501b-4da5-bf1e-4f4af8a04096" />

**Comment by tortoise10h** [comment]
Not sure if it's in the right thread but in Wezterm we have the option that `weight = 'DemiBold'` it's perfect for me since `Bold` is too bold. Same as current Ghostty, `font-thicken` only has 2 options `true` or `false`. Turn on the `font-thicken` seems too bold to me, I wish we had an option like Wezterm

**Reply by jcollie** [reply]
It's already there:

```
# The named font style to use for each of the requested terminal font styles.
# This looks up the style based on the font style string advertised by the
# font itself. For example, "Iosevka Heavy" has a style of "Heavy".
#
# You can also use these fields to completely disable a font style. If you set
# the value of the configuration below to literal `false` then that font style
# will be disabled. If the running program in the terminal requests a disabled
# font style, the regular font style will be used instead.
#
# These are only valid if its corresponding font-family is also specified. If
# no font-family is specified, then the font-style is ignored unless you're
# disabling the font style.
font-style = default
font-style-bold = default
font-style-italic = default
font-style-bold-italic = default
```

**Reply by oleksiivelychko** [reply]
Yes, it's possible here. I use these settings:
```
font-thicken = true
font-family = JetBrains Mono
font-style = Light
```
... and these also suit me too:
```
font-thicken = false
font-family = JetBrains Mono
font-style = SemiBold
```

**Reply by vitvly** [reply]
I've tried both
```
font-thicken = true
font-style = Light
```
and
```
font-thicken = false
font-style = SemiBold
```
but in both cases the result for Menlo 12 pt is still heavier than in Terminal.app.

**Comment by 00-kat** [comment, marked as answer]
Fixed with #4531, see the new config option [`font-thicken-strength`].

[`font-thicken-strength`]: https://ghostty.org/docs/config/reference#font-thicken-strength

**Comment by vitvly** [comment]
@00-kat thank you. However, i still don’t understand how to solve my original concern:
> how can i make font rendering completely identical to Terminal.app?

Which values of font-thicken-strength should i choose?

**Reply by 00-kat** [reply]
Apologies, I can't help with that as I don't use macOS and hence don't have a copy of Terminal.app to play with; you could try binary search a value for `font-thicken-strength` perhaps, or hopefully some other macOS user who happened to already configure this stumbles upon this post.

**Reply by matthewtodd** [reply]
@vitvly here's [what I ended up with](https://github.com/matthewtodd/dotfiles/commit/e7f58c926c3edfdf1d98e01f370625ab3a51e7e7), Menlo 12 like you.

**Reply by vitvly** [reply]
@00-kat @matthewtodd thank you! This looks much better.

I wonder though, does this mean that stock Apple apps like TextEdit or Terminal use some hidden font manipulations that third-party apps cannot fully replicate? Would be great if instead of imprecisely arriving at some values that look visually similar, we could be sure that we have it *exactly* the same.

**Comment by yaoyhu** [comment]
Anyupdate on it? It seems that font-thicken-strength is not working for me.

<img width="1512" alt="image" src="https://github.com/user-attachments/assets/4ffb7f3a-5ad5-433a-aae7-8e107966eb01" />

Additionally, I want to point out that Ghostty renders the same font differently, such as Fira Code. I had to switch to Fira Code Nerd Font to make the `>` symbol appear the same as it does in Kitty.

**Reply by 00-kat** [reply]
> Anyupdate on it?

Yeah, there's `font-thicken-strength` now.

> It seems that font-thicken-strength is not working for me.

In what way?  The two terminals in your screenshot look about the same (font thickness wise) to me.

> Additionally, I want to point out that Ghostty renders the same font differently, such as Fira Code. I had to switch to Fira Code Nerd Font to make the `>` symbol appear the same as it does in Kitty.

Open a new discussion.  In the new discussion make sure to actually send a comparison of the `>` glyphs.

**Reply by yaoyhu** [reply]
> In what way?

I believe this could make it appear clearer. Ghostty(left) looks darker and slimmer in contrast, while Kitty appears whiter and thicker.

<img width="500" alt="image" src="https://github.com/user-attachments/assets/0d78467b-6baf-4d5d-a64a-65b9d327743a" />


```py
# config in ghostty
font-size=17
font-family=FiraCode Nerd Font Mono
font-family=LXGW WenKai
font-thicken = true
font-thicken-strength = 228  # <--- No matter how I adjust this value and reload, it doesn't have any effect.
adjust-cell-width = "-2%"
adjust-cell-height = "6%"

# config in kitty
font_size 17
font_family      family="Fira Code"
bold_font        auto
italic_font      auto
bold_italic_font auto
macos_thicken_font 0.50
```

**Reply by 00-kat** [reply]
> No matter how I adjust this value and reload, it doesn't have any effect.

Oh also could you try restarting Ghostty?  It's possible that this is one of those configuration options that can't be hot-reloaded.  (You could also try the key you have bound to `reset_font_size`; `ghostty +list-keybinds` if you don't know which).

**Reply by yaoyhu** [reply]
I just changed the value abruptly from 0 to 255 and noticed a very slight difference. 

However, my Kitty font thickness is only set to 0.5, which is already **thicker** than ghossty's 255. Is there a way to make it even thicker?

**Reply by gelfand** [reply]
@yaoyhu Hey, have you found a solution?

**Reply by yaoyhu** [reply]
> @yaoyhu Hey, have you found a solution?

Nope, switching back to Kitty for now. Waiting for the fixed release.

**Reply by simplefr4nk** [reply]
I wanted to use the patched version of the `Rec Mono Casual`, but using `ghostty +list-fonts` didn't show it, it showed only the original one so I went with that.

I encountered the same problem with the font being bold instead of regular, and after trying the configurations above, nothing changed...

After using the Font Book application I saw that the font I was using had only three types; `bold`, `italic` and `bold-italic`. 

Switched to the patched one (that wasn't seen in the output of `ghostty +list-fonts`) with a `regular` version resolved everything, even without the `font-thicken-strength` configuration.

Maybe the font you're trying to use doesn't have a `regular` version?

**Reply by 00-kat** [reply]
Fira Code definitely does.
